posthog/special_migrations/migrations/events_sample_by_migration.py

```diff
@@ -0,0 +1,103 @@
+from constance import config
+
+from ee.clickhouse.client import sync_execute
+from ee.clickhouse.sql.events import EVENTS_TABLE, EVENTS_TABLE_MV_SQL, EVENTS_TABLE_SQL
+from posthog.constants import AnalyticsDBMS
+from posthog.settings import CLICKHOUSE_CLUSTER, CLICKHOUSE_DATABASE
+from posthog.special_migrations.definition import SpecialMigrationDefinition, SpecialMigrationOperation
+from posthog.version_requirement import ServiceVersionRequirement
+
+ONE_DAY = 60 * 60 * 24
+
+TEMPORARY_TABLE_NAME = "temp_events"
+
+
+class Migration(SpecialMigrationDefinition):
+
+    description = "Events table migration for compatible sample by column."
+
+    posthog_min_version = "1.30.0"
+    posthog_max_version = "1.31.0"
+
+    service_version_requirements = [
+        ServiceVersionRequirement(service="clickhouse", supported_version=">=21.6.0,<21.7.0"),
+    ]
+
+    operations = [
+        SpecialMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"DETACH TABLE {EVENTS_TABLE_MV_SQL} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            rollback=f"ATTACH TABLE {EVENTS_TABLE_MV_SQL} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            side_effect=lambda: setattr(config, "MATERIALIZED_COLUMNS_ENABLED", False),
+        ),
+        SpecialMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"CREATE TABLE IF NOT EXISTS {TEMPORARY_TABLE_NAME} AS {EVENTS_TABLE} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            rollback=f"DROP TABLE IF EXISTS {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+        ),
+        SpecialMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"""
+            INSERT INTO {TEMPORARY_TABLE_NAME}
+            SELECT * 
```

Comment:
Let's make this a "resumable" migration. You can do this by setting `resumable=True` and writing it in a way that can pick up from where it left off if it stops. 

from a comment on resumable ops:

```python
#   - Not resumable: `INSERT INTO table1 (col1) SELECT col1 FROM table2`
#   - Resumable: `INSERT INTO table2 (foo, timestamp) SELECT foo, timestamp FROM table1 WHERE timestamp > (SELECT max(timestamp) FROM table2)
```

Reply:
Should be safe to do a `>=` on max here because CH will dedupe the events anyway